### PR TITLE
Improve FQSEN support for @see

### PR DESCRIPTION
**Pull request by mikehaertl**

### CODE REVIEW

File: corpus/references.txt [hunk 1]
```diff
@@ -146,24 +145,376 @@
...
+ * @see \MyClass::some_function()
+ * @see \Vendor\Package\ClassName::some_function()
+ * @see MyClass::some_function()
+ * @see Vendor\Package\ClassName::some_function()
  */
 ---
 (document
+  (tag
+    (tag_name)
+    (fqsen (name)))
```
Comment (mikehaertl) [inline corpus/references.txt:202]: I'm unsure if `fqsen` is really appropriate here. I would expect that it always includes a full class name like in `Vendor\Package\ClassName::$someVar`. But then even in the examples on phpdoc.org they don't include the class path.

File: grammar.js [hunk 2]
```diff
@@ -498,7 +496,16 @@ module.exports = grammar({
...
-    uri: $ => /\w+:(\/?\/?)[^\s}]+/,
+    uri: $ => /https?:\/\/[^\s}]+/,
+
+    fqsen: $ => choice(
+      $.named_type,
+      $.variable_name,
+      seq($.name, '()'),
+      seq($.named_type, '::', $.name),
+      seq($.named_type, '::', $.name, '()'),
+      seq($.named_type, '::', $.variable_name)
```
Comment (mikehaertl) [inline grammar.js:507]: Here I wondered if we should reuse things like [scoped_property_access_expression](https://github.com/tree-sitter/tree-sitter-php/blob/0ce134234214427b6aeb2735e93a307881c6cd6f/grammar.js#L1027-L1031) or [class_constant_access_expression](https://github.com/tree-sitter/tree-sitter-php/blob/0ce134234214427b6aeb2735e93a307881c6cd6f/grammar.js#L882-L886). But this would result in a quite different node structure. And in our case it's not really an access either. So I went with a definition similar to types.

File: corpus/references.txt [hunk 3]
```diff
@@ -146,24 +145,376 @@
...
+ */
+---
+(document
+  (tag
+    (tag_name)
+    (fqsen
+      (named_type
+        (qualified_name
+          (namespace_name_as_prefix)
+          (name))))
```
Comment (mikehaertl) [inline corpus/references.txt:343]: For all these final `(name)` nodes I'm also unsure. They either represent:

 * a class name (like here)
 * a method name
 * a property name
 * a constant name

We will not be able to distinguish between these in a query later. And we also can't e.g. query for the property name explicitely as `(name)` is so generic.

**Comment by mikehaertl** [comment]
@claytonrcarter Bump :)

File: grammar.js (hunk 2 above)
Comment (claytonrcarter) [reply grammar.js:507]: So, I *think* that I would prefer to explicitly share/reuse the PHP parser nodes in this case. The upside of doing so, in my limited experience, is that it means that we can have grammar queries that match the same syntax "things" whether they appear in code or in docblocks, which has been nice. eg by reusing the type definitions where we parse types for return/param/etc has meant that my editor has started making those clickable w/o any add'l config or effort, ie just by happy accident. Reusing the php parser nodes here might also make the same work for `@see`. 

Unless I'm mistaken, the 3 class access nodes we're looking for are `class_constant_access_expression`, `scoped_property_access_expression` and `scoped_call_expression`. I wonder if `function_call_expression` would cover the bases for L504, or if something else might be needed there.

Please advise if you think going this route would be unwise, though. Perhaps you can elaborate more on why you went this route vs reusing PHP nodes?

File: corpus/references.txt (hunk 3 above)
Comment (claytonrcarter) [reply corpus/references.txt:343]: I think that -- where we're parsing things that are really just PHP syntax -- sticking as close to the PHP syntax parser/node structure as possible would be best. So as long as we're using `(name)` like they use `(name)`, then I think we'll be OK.

**Comment by claytonrcarter** [comment]
@mikehaertl Thanks for the nudge and thank you for your ongoing contributions and I'm really glad that you're find this useful, and helping to make it more useful!

Can you tell me more about how how you intend to use the `(fqsen)` node type? I had missed that when reviewing #9 and I'm a little bit 🤔 wondering if that node should instead be a hidden node (ie `$._fqsen`) just to help us in building the parser instead of a regular named node. Have you come across a usecase where you wanted to query for *any* FQSEN vs for a specific type of FQSEN? (I guess that gets back to the question of reusing the node types from the main parser so consumers could reuse the same queries in PHP as they use in phpDoc to find, eg `scoped_call_expression`, etc).

Again, I'm mostly wondering aloud and looking for thoughts and input; not set in stone.

Thanks again!

**Comment by mikehaertl** [comment]
> Can you tell me more about how how you intend to use the (fqsen) node type?

Well it's part of the (semi-)official phpdoc specs like in `@see [URI | FQSEN] [<description>]`. We have a `uri` node so we should also have `fqsen` node. After all we build the *phpdoc* parser so IMO it's not surprising that our node structure may be a little different than PHP's. And PHP's node structure may not yet be set in stone either.

Regarding reusing PHP parser nodes: I see all your points and they make sense. But on the other hand I also have some doubts:

1. We can't really reuse rules as they are. Instead we go deep into the internals of the PHP grammar and "cobble together" our own modified rules from them like here:
https://github.com/claytonrcarter/tree-sitter-phpdoc/blob/8fae507a8d749f58958ba160f747231990fb55f0/grammar.js#L439-L446

    This gives me doubts if e.g. upgrades are really painless later (which is one of the promises of code reuse). We will _always_ have to carefully inspect the updates in the PHP grammar and which of them are relevant. We may also have to pick them apart like above.

2. In addition I found having to look at two grammar files makes things hard to follow if you need to dig into a rule. You need to be very focussed to not get lost. The heavyweight PHP grammar doesn't help in that respect.

These points made me wonder if the hassle of importing the PHP grammar is really worth it or if we shouldn't simply *copy* the few relevant rules into our grammar file. It would make our file so much easier to digest.

Also just thinking aloud here. Thoughts?

File: grammar.js (hunk 2 above)
Comment (mikehaertl) [reply grammar.js:507]: The problem I see here is that the PHP rules match more than what we need here and thus are not really apropriate. For example `scoped_property_access_expression` uses `_scope_resolution_qualifier` which uses  [_dereferencable_expression](https://github.com/tree-sitter/tree-sitter-php/blob/0ce134234214427b6aeb2735e93a307881c6cd6f/grammar.js#L1150-L1159). I think it covers expressions like `$className::someMethod()` which can not appear in a docblock. Like I said: Syntactically we don't deal with accessing a prop/constant here which has way more options than what we need here.

See my other answer below for my thoughts on PHP rule reuse.


> we can have grammar queries that match the same syntax "things" whether they appear in code or in docblocks
> [...]  my editor has started making those clickable w/o any add'l config or effort, ie just by happy accident

But isn't this up to us, which queries we use for phpdoc? We don't have to reuse the PHP queries. Thinking more about it, are you sure that this happy accident is really due to treesitter? At least in nvim this is covered by LSP and treesitter is only used for syntax highlighting.

**Comment by claytonrcarter** [comment]
Sorry for the long cycle times on this discussion. TBH I'm still no settled on this, but I don't want my vacillation to block any more progress for you. This is clearly well thought out and executed. We can discuss the parser approach separately and make revisions if needed. Thank you!

**Comment by claytonrcarter** [comment]
> Well it's part of the (semi-)official phpdoc specs

Yup, fair enough. I knew that, but just needed to be pushed over the edge. 😄 

> are you sure that this happy accident is really due to treesitter?

Not at all! 😄 

For all other points, I'm going to create an issue for us to discuss the copy/paste vs import/compose discussion, and how to structure the parser.
